streams/src/main/java/org/apache/kafka/streams/kstream/Consumed.java

```diff
@@ -129,7 +137,19 @@ protected Consumed(final Consumed<K, V> consumed) {
      * @return a new instance of {@link Consumed}
      */
     public static <K, V> Consumed<K, V> with(final Topology.AutoOffsetReset resetPolicy) {
-        return new Consumed<>(null, null, null, resetPolicy);
+        return new Consumed<>(null, null, null, resetPolicy, null);
+    }
+
+    /**
+     * Create an instance of {@link Consumed} with provided processor name.
+     *
+     * @param processorName the processor name to be used. If {@code null} a default processor name will be generated
+     * @param <K>         key type
+     * @param <V>         value type
+     * @return a new instance of {@link Consumed}
+     */
+    public static <K, V> Consumed<K, V> with(final String processorName) {
```

Comment:
Personally I like to rename it (and others on the other classes) to `as`. I know this is a subjective preference and hence would like to hear more thoughts? @fhussonnois @mjsax @vvcephei @bbejeck

Reply:
if we rename to `as` we should also change the Suppressed interface to rename `withName` to `as` for consistency ?

Reply:
Good call. Actually there are also `Grouped` and `Joined` which are already released as part of 2.1.

So how about renaming it to `withName` instead of `with` for consistency with Grouped, Joined, and Suppresed? cc @vvcephei @bbejeck

Reply:
Actually : 
-  `Grouped` class has a static method `as` and instance method `withName`
- `Joined` class has a static method  `named` and instance method `withName`
- `Suppressed` class has no static method to defined name and instance method `withName`

So it's OK to rename the static methods `with` to `as` but we should change the `Joined` class too.

Reply:
I think this is a good point. It is a regret that 2.1 is already released with `Joined.named` but personally I think it still makes more sense to predicate and replace it with `Joined.as`, while @fhussonnois get all others to be consistent with `as`. @bbejeck wdyt?

Reply:
Hey @fhussonnois @vvcephei @bbejeck  I've made one more pass over all the control classes we have so far, and here is a list along with the current state method name:

0. Named: `as` (new, Aggregator / Mapper / Predicate / Foreach / Joiner / and SuppressedInternal extends it)
1. Produced / Consumed: `with` (new)
2. Printed: `withName` (new)
3. Grouped: `as`, BUT this is used for repartition topic names only
4. Joined: `named`, BUT this is used for repartition topic names only
5. Suppressed: `withName`, BUT this is used for both processor name and store name
6. Materialized: `as`, BUT this is used for state store names only

So as I read it, the semantics of 3/4/6 are not correlated to 0/1/2, as they are used for naming the processors only, not repartition topics / state stores.

In my other comment I was wondering if we could have 1/2 extending 0 as well, because the operators leveraging them do not have a `Aggregator / etc` lambda that extends Named so we are sort of using those three as a "shortcut" to still allow users to override names. For suppression operator, it is the similar rationale (i.e. no transformation and hence no lambda functions, just leverage on the control object). Following this, I think 5 can also extend 0 as well (in your current impl only `SuppressedInternal` extends `NamedInternal`, I'd suggest we let `Named` -> `Suppressed` -> `SuppresedInternal` and moving `Named.processorName()` as package private, same as 1/2). 

-------------------------------------

So I'd suggest let Produced / Consumed / Printed / Suppressed to extend Named. Move `processorName` to Named as package private so that all sub-classes can use. Keep `Named.as` function name across these functions as well because `Named.withName` or `Named.named` sounds quite awkward.

-------------------------------------

For 6, a.k.a. `Materialized.as` is really for a different scope and orthogonal to this discussion, so I'd suggest we keep it as-is. 

-------------------------------------

For 3/4, this is a bit tricky since now we are effectively providing multiple ways to allow users to set repartition topics. To be more concrete, for example:

a) `stream1.join(stream2, named("name1", /* join logic */))` -> use `name1` for processor, and also `name1-repartition` as the repartition topic.

b) `stream1.join(stream2, /* join logic*/, Joined.named("name2"))` -> use `name2` for repartition topic, internal name for processor.

c) `stream1.join(stream2, named("name1", /* join logic */), Joined.named("name2"))` -> use `name1` for processor name, and `name2-repartition` for repartition name.

This sounds to me a bit too confusing to users, and hence makes me wondering if we should just keep one of the two set of APIs. WDYT?